### DISCUSSION TITLE
Getting Real-Time output when running salesforce cli command

### DISCUSSION THREAD

**Discussion by StreckerCM** [Tyrrrz/CliWrap#214] [Q&A]
I have an app that uses the salesforce cli and I want to send what you would get when running the command manually in the Powershell window to a textbox on the screen.  I call the command using observe, but the only event that gets triggered is started or stdErr.  I never see a StdOut command trigger, but I can open the log file for the command in notepad++ and see that it is running.

My only thought is it has something to do with how Salesforce CLI is writing to the console, but I need to figure out how to solve it.  I tried doing a file system watcher event on the log file, but that causes the command to fail even if I open it in read-only.

**The output when running the command in powershell**
![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/351f33da-6b49-47e6-9a36-e83ba5ec2618)

**My CliWrap helper class**
```
using CliWrap;
using CliWrap.EventStream;
using Salesforce_Sandbox_Utility;
using System;
using System.Diagnostics;
using System.IO;
using System.Reactive.Linq;
using System.Threading.Tasks;

namespace SFDC_Sandbox_Utility.Classes
{
	public delegate void OnStart(int processId);
	public delegate void OnExit(int exitCode);
	public delegate void OnTextStreamHandler(string text);

	public class CliWrapHelper
	{
		public int threadID { get; set; }

		public event OnTextStreamHandler? OnStdOutput;
		public event OnTextStreamHandler? OnStdErr;
		public event OnStart? OnStart;
		public event OnExit? OnExit;

		public async Task ExecutCommand()
		{
			Command = Cli.Wrap("powershell")
						.WithWorkingDirectory(Paths.sfdx)
						.WithArguments(a => a
							.Add("sfdx")
							.Add("sfdmu:run")
							.Add("-s")
							.Add(sourceOrg)
							.Add("-u")
							.Add(item)
							.Add("-p")
							.Add(Path.GetDirectoryName(scriptRunPath))
							.Add("--noprompt")
							.Add("--verbose"))

			await inCommand.Command.Observe().ForEachAsync(cmdEvent =>
			{
				switch (cmdEvent)
				{
					case StartedCommandEvent started:
						threadID = started.ProcessId;
						OnStart?.Invoke(started.ProcessId);
						Debug.WriteLine($"Process started; ID: {started.ProcessId}");
						break;
					case StandardOutputCommandEvent stdOut:
						OnStdOutput?.Invoke(stdOut.Text);
						Debug.WriteLine($"Out> {stdOut.Text}");
						break;
					case StandardErrorCommandEvent stdErr:
						OnStdErr?.Invoke(stdErr.Text);
						Debug.WriteLine($"Err> {stdErr.Text}");
						break;
					case ExitedCommandEvent exited:
						OnExit?.Invoke(exited.ExitCode)
						Debug.WriteLine($"Process exited; Code: {exited.ExitCode}");
						break;
					default:
						Debug.WriteLine(cmdEvent.ToString());
						break;
				}
			});

		}

	}
}
```

**Comment by Tyrrrz** [comment]
Just to test, if you run `ExecuteBufferedAsync()` (instead of `Observe()`) is the stdout captured there?

**Reply by StreckerCM** [reply]
Sorry for the delay. I wasn't able to get to it over the weekend. I ran it using `ExecuteBufferedAsync()` and got the same output as when running it in Powershell.

![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/842ccb95-c6f7-47e7-9a33-fcca0b8d23ce)

**Reply by Tyrrrz** [reply]
Interesting. What if you use `ListenAsync()` instead? It's slightly different than `Observe()` but operates based on the same backing model. Trying to isolate the issue.

**Reply by StreckerCM** [reply]
I tried it with `ListenAsync()` and monitored the process using Notepad++ to watch the log file.  I got the process started event, but all the standard out events waited until after the process had been completed.

I did a screen record; if you jump to the end, you will see the std output dump into the immediate window just before the process exited event https://youtu.be/ZhLBYFGDBe4

**Reply by Tyrrrz** [reply]
I wonder if the terminal codes might be interfering somehow 🤔 

Okay, final test, does this work?

```csharp
var target = PipeTarget.Create(async (stream, cancellationToken) =>
{
    using var reader = new StreamReader(stream);
    var buffer = new char[128];

    while (true)
    {
        var charsRead = await reader.ReadAsync(buffer, 0, buffer.Length, cancellationToken);
        if (charsRead <= 0)
            break;

        var text = new string(buffer, 0, charsRead);
        Debug.Write(text);
    }
});

var cmd = Cli.Wrap("powershell")
	.WithWorkingDirectory(Paths.sfdx)
	.WithArguments(a => a
		.Add("sfdx")
		.Add("sfdmu:run")
		.Add("-s")
		.Add(sourceOrg)
		.Add("-u")
		.Add(item)
		.Add("-p")
		.Add(Path.GetDirectoryName(scriptRunPath))
		.Add("--noprompt")
		.Add("--verbose"))
    .WithStandardOutputPipe(target)
    .WithStandardErrorPipe(target);
```

**Reply by StreckerCM** [reply]
I had to tweak it slightly to get VS to accept the code.  I am using CliWrap from NuGet version 3.6.4, and the error said that there isn't a `ReadAsync()` that takes four arguments.

![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/6b3958ee-f2e3-47f2-9b71-fbfac2fb312e)

These are the two options it showed for `ReadAsync()`

![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/df2f4fa6-a604-4a24-93d8-ff6b5700888f)

![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/67cb1023-fb29-48b1-a650-11dcec3bd733)

So I removed the `cancellationToken` argument to get it to compile.  I still have a warning that the method lacks an `await` operator, but it is in the pipeline target structure.  I also had to add `ExecuteAsysn()` because the command wouldn't start without it.  Let me know if I should tweak any of these changes and test again.

![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/bc169d2e-9f8d-4e3d-810d-6c873e1c6872)

However, it looks like it does the same as when I ran it all the other times.  During the process, I see "Command in progress..." Once it completes, it dumps it all into the immediate window.

![image](https://github.com/Tyrrrz/CliWrap/assets/3030358/6fe5c281-1464-48ed-82f6-08b19d6e1ba3)

**Reply by Tyrrrz** [reply]
Sorry, yeah. You should put `await` before `ExecuteAsync()`.

I suspect the program never flushes the data when the output is redirected, for some reason. Can you create a minimal repro zip/repository so I can test it myself?

**Reply by StreckerCM** [reply]
@Tyrrrz I created a repo with a sample console app that will use the salesforce CLI to connect to a playground org I created and load some mock data.  The readme file should actually have everything you need to set up the environment but don't hesitate to let me know if you have any issues.  Good luck!

**Reply by Tyrrrz** [reply]
Thank you @StreckerCM. I will take a look when I have a bit more free time.

**Reply by Tyrrrz** [reply]
Hey @StreckerCM. I tried running the program and it seems I'm getting the following error (taken from the log file):

```
 [20:22:20.157] Execution of the command sfdmu:run has been started.

 LOADING AND VALIDATING THE EXPORT.JSON SCRIPT ...

 OBJECT SET #1 STARTED

 {
   "Running Plugin version": "4.29.7"
}
 {
   "Running Add-On API version": "1.5.0"
}

 [20:22:20.171] Loading Core Add-On Module declarations ...
 [20:22:20.222] Connecting to the CliWrap using SFDX force:org:display ...

 [20:22:21.654] [ERROR] Execution of the command has aborted due to unexpected error: Command failed: sfdx force:org:display --json --targetusername CliWrap.
 [20:22:21.656] [ERROR] [STACK TRACE] Error: Command failed: sfdx force:org:display --json --targetusername CliWrap
    at checkExecSyncError (node:child_process:885:11)
    at execSync (node:child_process:957:15)
    at Common.execSfdx (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\modules\components\common_components\common.js:196:49)
    at ScriptOrg._setupConnection (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\modules\models\script_models\scriptOrg.js:124:53)
    at ScriptOrg.setupAsync (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\modules\models\script_models\scriptOrg.js:52:20)
    at Script.setupAsync (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\modules\models\script_models\script.js:213:30)
    at async RunCommand.setupObjectSetAsync (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\modules\commands_processors\runCommand.js:71:9)
    at async RunCommandExecutor.execute (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\modules\commands_processors\runCommandExecutor.js:39:17)
    at async Run.run (C:\Users\Tyrrr\AppData\Local\sf\node_modules\sfdmu\lib\commands\sfdmu\run.js:20:9)
 [20:22:21.658] [ERROR] Execution of the command sfdmu:run has been completed. Exit code 1 (COMMAND_UNEXPECTED_ERROR).
 [20:22:21.659] Total time elapsed: 00h 00m 01s 500ms .
```

However, the output seems to be reproduced the same way, whether I'm running the script manually or through the program.

While reading your code, the only thing that stood out (that I didn't originally notice) was that you're wrapping `powershell` for some reason, instead of `sfdx`. Try replacing `Cli.Wrap("powershell")` with `Cli.Wrap("sfdx")`.

**Comment by bao86** [comment]
Your solution is working fine thanks :)

Do you know how to Encoding UTF8 With BOM?

![image](https://github.com/Tyrrrz/CliWrap/assets/103753661/81444a84-cd13-4c9e-96e6-19adde2949f9)

**Reply by Tyrrrz** [reply]
If the program is producing valid UTF-8 content, then you can just change the encoding on CliWrap's side using one of the overloads.

**Comment by bao86** [comment]
This is my Powershell script :
```
[CmdletBinding()]
param(
    [parameter(Mandatory = $false)]
	[ValidateNotNullOrEmpty()]
	[string]$Marque,

    [parameter(Mandatory = $false)]
	[ValidateNotNullOrEmpty()]
	[string]$Modele,

    [parameter(Mandatory = $false)]
	[ValidateNotNullOrEmpty()]
	[string]$Couleur,

    [parameter(Mandatory = $false)]
	[ValidateNotNullOrEmpty()]
	[string]$LogName
)

Write-Output "INFO:Marque : $Marque"
Write-Output "INFO:Modele : $Modele"
switch ($Couleur)
{
    "Rouge" 
    {
        Write-Output "ERROR:Couleur sélectionné: $Couleur"
    }
    "Orange"
    {
        Write-Output "WARNING:Couleur sélectionné: $Couleur"
    }
    Default
    {
        Write-Output "INFO:Couleur sélectionné: $Couleur"
    }
}
```

C# : 
```
private async Task Execute(object sender, RoutedEventArgs e) 
{
            [code] blablabla [code]

            var target = PipeTarget.Create(async (stream, cancellationToken) =>
            {
                using var reader = new StreamReader(stream);
                var buffer = new char[8192];

                while (true)
                {
                    var charsRead = await reader.ReadAsync(buffer, 0, buffer.Length);

                    if (charsRead <= 0)
                        break;
                    var text = new string(buffer, 0, charsRead);

                    string message = UnicodeToUTF8(text);
                    Debug.Write("!DEBUT! " + message + " !FIN!");
  
                }
            });

            var psResults = await Cli.Wrap("powershell")
                    .WithArguments(args => args
                        .Add(values))
                    .WithValidation(CommandResultValidation.None)
                    .WithStandardOutputPipe(target)
                    .WithStandardErrorPipe(target)
                    .ExecuteBufferedAsync();

}

static string UnicodeToUTF8(string from)
{
            var bytes = Encoding.UTF8.GetBytes(from);
            return new string(bytes.Select(b => (char)b).ToArray());
}

static string ConvertStringToUtf8Bom(string source)
{
            var data = Encoding.Default.GetBytes(source);
            var result = Encoding.UTF8.GetPreamble().Concat(data).ToArray();
            var encoder = new UTF8Encoding(true);

            return encoder.GetString(result);
}
```

Result when using **UnicodeToUTF8** method: 
!DEBUT! INFO:Marque : Sï¿½at !FIN!!
!DEBUT! INFO:Modele : Atï¿½cï¿½ !FIN!!
!DEBUT! INFO:Couleur sï¿½lectionnï¿½: Noir !FIN!!

Result when using **ConvertStringToUtf8Bom** method: 
!DEBUT! ﻿INFO:Marque : S�at !FIN!!
!DEBUT! ﻿INFO:Modele : At�c� !FIN!!
!DEBUT! ﻿INFO:Couleur s�lectionn�: Noir !FIN!!

Result expected : 
!DEBUT! ﻿INFO:Marque : Séat !FIN!!
!DEBUT! ﻿INFO:Modele : Atécà !FIN!!
!DEBUT! ﻿INFO:Couleur sélectionné: Noir !FIN!!

I don't know how to convert :/

**Reply by Tyrrrz** [reply]
```csharp
using var reader = new StreamReader(stream, Encoding.UTF8);
```

**Reply by bao86** [reply]
Already test and same result :(

**Reply by bao86** [reply]
I try this : 
```
var target = PipeTarget.Create(async (stream, cancellationToken) =>
{
	var enc = Encoding.UTF8;
	using var reader = new StreamReader(stream, enc);
	var testm = await reader.ReadToEndAsync();
	MessageBox.Show(testm);

	[code]blablabla[/code]
}
```
![image](https://github.com/Tyrrrz/CliWrap/assets/103753661/055dffe9-0a25-495f-a685-6c01ae393f5e)

also with this : 
`.ExecuteBufferedAsync(Encoding.UTF8);`
same result

**Reply by Tyrrrz** [reply]
You might have to switch the terminal itself into unicode mode using `chcp`. Search around, there was a discussion/issue about this.

**Reply by bao86** [reply]
It's working with your extension method 👍 
Thank you !!

https://github.com/Tyrrrz/CliWrap/issues/196#issuecomment-1445190176
